handle None namespace and version suffix in conversion

fxbuson opened the pull request:
As mentioned in Issue #283, the current version of the converter doesn't know how to handle identities that end in the version number. You get these in the case of using `sbol2.ConfigOptions.SBOL_COMPLIANT_URIS` as `True`, so I added a function `_sbol3_identity()` to be used whenever converting an identity value from sbol2 to sbol3, and exchanged each case of an sbol3 object being assigned an identity to use that function.

The converter also didn't include the version number in the identity when converting back to sbol2, so I included that in `_convert_identified()`. Note that this will default to '1', as per the SBOL 3.1.0 specification, section 10.2, item 3:

> When SBOL 3.x Identified object is converted to SBOL 2.x, if its identity is a URL, then the identity of 
the SBOL 2.x object SHOULD be constructed as [SBOL3 identity]/[SBOL2 version]. If the object does
not have an SBOL2 version property, then its version SHOULD default to 1. 

Another problem I ran into was with `None` namespaces when converting to sbol3, so I added that case to `_sbol3_namespace()`.

Now, my conversion of a simple document with an example ComponentDefinition should work. The roundtrip will enforce the use of the version at the end of the object's identity though, overriding the option `sbol2.ConfigOptions.SBOL_COMPLIANT_URIS` if it was set to `False`

fxbuson committed:
handle None namespace and version suffix in conversion

Created a function that strips the end of the identity when converting to sbol3 in case it ends with the version number (as per the sbol2.3 specification)

Added this conversion to all implemented objects

Added roundtrip inclusion of sbol2version in the sbol2 identity. Defaults to "1"

PrashantVaidyanathan commented:
@jakebeal @tcmitchell - Just wanted to tag you in this. Are you happy with this?

fxbuson committed:
fix rstrip misunderstanding

also properly determining identities in toplevel objects

jakebeal commented on sbol_utilities/sbol3_sbol2_conversion.py:
```diff
@@ -73,6 +73,10 @@ def _convert_identified(self, obj3: sbol3.Identified, obj2: sbol2.Identified):
         obj2.wasGeneratedBy = obj3.generated_by
+
+        # Add version to identity
+        obj2.identity = obj2.identity + "/" + obj3.sbol2_version
```
I don't think it's safe to assume that the `sbol2_version` property is necessarily set.

jakebeal commented on sbol_utilities/sbol3_sbol2_conversion.py:
```diff
@@ -341,6 +345,11 @@ def _convert_toplevel(self, obj2: sbol2.TopLevel, obj3: sbol3.TopLevel):
         obj3.attachments = [a.identity for a in obj2.attachments]
 
+    def _sbol3_identity(self, obj2: sbol2.Identified):
+        if obj2.identity.endswith("/" + obj2.version):
```
```suggestion
        if obj2.version and obj2.identity.endswith("/" + obj2.version):
```
The `obj2.version` property is not required to be set. If this is the case, then this conversion might not work correctly.

jakebeal reviewed (changes requested):
This also needs to have the GitHub action tests run on it (turn them on in your fork) and needs to have new test cases added that are capable of reproducing the failures in #283 if these changes are not added.

fxbuson committed:
check if sbol3 object has sbol2 version

fxbuson committed:
handle case without sbol2 version

Co-authored-by: Jacob Beal <jakebeal@gmail.com>

fxbuson committed:
Merge branch 'issue_283' of https://github.com/fxbuson/SBOL-utilities into issue_283

fxbuson committed:
update CI workflow

dropped python versions 3.7 and 3.8 (reaching EOL by the end of the year) and adding 3.12

fxbuson committed:
add simplest pysbol2 CD test file

as described in issue 283

fxbuson committed:
add sbol2 version check

fxbuson committed:
Merge pull request #1 from SynBioDex/develop

merge latest develop branch with CI corrections

fxbuson committed:
Merge pull request #2 from SynBioDex/develop

merge latest develop branch with CI corrections
